crates/compilers/src/lib.rs

```diff
@@ -882,6 +890,37 @@ fn rebase_path(base: &Path, path: &Path) -> PathBuf {
     new_path.to_slash_lossy().into_owned().into()
 }
 
+/// Utility function to apply a set of updates to provided sources.
+fn apply_updates(sources: &mut Sources, updates: Updates) {
+    for (path, source) in sources {
+        if let Some(updates) = updates.get(path) {
+            source.content = Arc::new(replace_source_content(
+                source.content.as_str(),
+                updates.iter().map(|(start, end, update)| ((*start..*end), update.as_str())),
+            ));
+        }
+    }
+}
+
+/// Utility function to change source content ranges with provided updates.
+/// Assumes that the updates are sorted.
+fn replace_source_content<'a>(
+    source: &str,
+    updates: impl IntoIterator<Item = (Range<usize>, &'a str)>,
+) -> String {
+    let mut offset = 0;
```

Comment:
I suspect the win panic is due to the offset not being applied / incremented for all sources but only for specific